xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,393 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package gracefulswitch implements a graceful switch load balancer.
+package gracefulswitch
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/resolver"
+)
+
+var (
+	errBalancerClosed = errors.New("gracefulSwitchBalancer is closed")
+	defaultState      = balancer.State{
+		ConnectivityState: connectivity.Connecting,
+		Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+	}
+)
+
+func newGracefulSwitchBalancer(cc balancer.ClientConn, opts balancer.BuildOptions) *gracefulSwitchBalancer {
+	return &gracefulSwitchBalancer{
+		cc:    cc,
+		bOpts: opts,
+	}
+}
+
+type gracefulSwitchBalancer struct {
+	bOpts balancer.BuildOptions
+	cc    balancer.ClientConn
+
+	mu              sync.Mutex // also guards all fields within balancerCurrent and balancerPending
+	balancerCurrent *balancerWrapper
+	balancerPending *balancerWrapper
+	closed          bool // set to true when this balancer is closed
+
+	// currentMu must be locked before mu.
+	currentMu sync.Mutex // protects operations on the balancerCurrent
```

Comment:
Do we still need this, actually, given how we update `balancerCurrent` and `balancerPending` and store the subconns inside them now?  `UpdateSubConnState` calls won't find the balancer we're closing if `swap` has been called.

Reply:
Argh, that's right.  `UpdateSubConnState` could be called first, find the child, then give up the lock, then `swap` could cause the close to happen before the method on the child is called.  It's an unlikely race, but possible.  Let's make sure this sequence is mentioned in the comments, then.

Reply:
Added this comment to currentMu: "This mutex guards against this sequence of events: UpdateSubConnState() called, finds the balancerCurrent, gives up lock, updateState comes in, causes Close() on balancerCurrent before the UpdateSubConnState is called on the balancerCurrent."